src/main/java/org/cryptomator/cryptolib/v3/FileNameCryptorImpl.java

```diff
@@ -0,0 +1,71 @@
+/*******************************************************************************
+ * Copyright (c) 2015, 2016 Sebastian Stenzel and others.
+ * This file is licensed under the terms of the MIT license.
+ * See the LICENSE.txt file for more info.
+ *
+ * Contributors:
+ *     Sebastian Stenzel - initial API and implementation
+ *******************************************************************************/
+package org.cryptomator.cryptolib.v3;
+
+import com.google.common.io.BaseEncoding;
+import org.cryptomator.cryptolib.api.AuthenticationFailedException;
+import org.cryptomator.cryptolib.api.FileNameCryptor;
+import org.cryptomator.cryptolib.api.Masterkey;
+import org.cryptomator.cryptolib.common.DestroyableSecretKey;
+import org.cryptomator.cryptolib.common.MessageDigestSupplier;
+import org.cryptomator.cryptolib.common.ObjectPool;
+import org.cryptomator.siv.SivMode;
+import org.cryptomator.siv.UnauthenticCiphertextException;
+
+import javax.crypto.IllegalBlockSizeException;
+import java.security.MessageDigest;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+class FileNameCryptorImpl implements FileNameCryptor {
+
+	private static final BaseEncoding BASE32 = BaseEncoding.base32();
+	private static final ObjectPool<SivMode> AES_SIV = new ObjectPool<>(SivMode::new);
+
+	private final Masterkey masterkey;
+
+	FileNameCryptorImpl(Masterkey masterkey) {
+		this.masterkey = masterkey;
+	}
+
+	@Override
+	public String hashDirectoryId(String cleartextDirectoryId) {
+		try (DestroyableSecretKey ek = masterkey.getEncKey(); DestroyableSecretKey mk = masterkey.getMacKey();
+			 ObjectPool.Lease<MessageDigest> sha1 = MessageDigestSupplier.SHA1.instance();
+			 ObjectPool.Lease<SivMode> siv = AES_SIV.get()) {
+			byte[] cleartextBytes = cleartextDirectoryId.getBytes(UTF_8);
+			byte[] encryptedBytes = siv.get().encrypt(ek, mk, cleartextBytes);
+			byte[] hashedBytes = sha1.get().digest(encryptedBytes);
+			return BASE32.encode(hashedBytes);
+		}
+	}
+
+	@Override
+	public String encryptFilename(BaseEncoding encoding, String cleartextName, byte[]... associatedData) {
+		try (DestroyableSecretKey ek = masterkey.getEncKey(); DestroyableSecretKey mk = masterkey.getMacKey();
+			 ObjectPool.Lease<SivMode> siv = AES_SIV.get()) {
+			byte[] cleartextBytes = cleartextName.getBytes(UTF_8);
+			byte[] encryptedBytes = siv.get().encrypt(ek, mk, cleartextBytes, associatedData);
```

Comment:
@overheadhunter from the overview picture in https://github.com/encryption-alliance/unified-vault-format/pull/19, I would have expected file name encryption not to depend on the `fileKey` (enc part of `masterkey`)?

Reply:
Right, this isn't correct yet. File name encryption is currently just copy/pasted from the older format.

Reply:
switched to derived keys:

https://github.com/cryptomator/cryptolib/blob/084e78a213d7fd5f193d8f545cd7a366988ba996/src/main/java/org/cryptomator/cryptolib/v3/FileNameCryptorImpl.java#L28-L40